libr/bin/p/bin_xnu_kernelcache.c

```diff
@@ -1882,6 +1902,214 @@ static RStubsInfo *get_stubs_info(struct MACH0_(obj_t) *mach0, ut64 paddr, RKern
+	const RVector *sections = MACH0_(load_sections) (kext->mach0);
+	if (!sections) {
+		return;
+	}
+
+	RList *classes = r_list_newf (r_iokit_class_free);
+
+	RBinSymbol *sym;
+	R_VEC_FOREACH (symbols, sym) {
+		if (!strstr (sym->name->oname, ".init."))
+			continue;
+		fprintf (stderr, "\n=== Parsing: %s\n", sym->name->oname);
+
+		ut64 func_vaddr = sym->vaddr;
+		ut64 reg_values[29] = { 0, };
+
+		for (ut64 pc = func_vaddr; true; pc += 4) {
+			int reg;
+			ut64 addr;
+			ut32 insn;
+			if (try_parse_adrp_add_pair (bf, kext, pc, &reg, &addr, &insn)) {
+				if (reg < R_ARRAY_SIZE (reg_values))
+					reg_values[reg] = addr;
+				continue;
+			}
+
+			int dst_reg, src_reg;
+			if (try_parse_mov_reg_reg (insn, &dst_reg, &src_reg)) {
+				if (dst_reg < R_ARRAY_SIZE (reg_values) &&
+						src_reg < R_ARRAY_SIZE (reg_values)) {
+					reg_values[dst_reg] = reg_values[src_reg];
+				}
+				continue;
+			}
+
+			ut32 imm16 = 0;
+			if (try_parse_movz_w3_imm (insn, &imm16)) {
+				reg_values[3] = imm16;
+				continue;
+			}
+
+			if (is_ret_like (insn))
+				break;
+
+			if (!is_bl (insn))
+				continue;
+
+			const int off26 = (int) ((insn & 0x03FFFFFFu) << 6) >> 6;
+			const ut64 bl_target = pc + ((ut64) off26 << 2);
+
+			if (reg_values[1] == 0)
+				continue;
+			char *name = NULL;
+			if (!try_read_printable_cstr (bf, kext, reg_values[1], &name))
+				continue;
+
+			RIOKitClass *c = R_NEW0 (RIOKitClass);
+			if (!c)
+				continue;
+			c->name = name;
+			c->size = reg_values[3];
+			c->meta_va = reg_values[0];
+			c->supermeta_va = reg_values[2];
+			r_list_append (classes, c);
+
+			fprintf (stderr, "Found IOKit class: %s\n", name);
+		}
+	}
+
+	RListIter *it;
+	RIOKitClass *c;
+	r_list_foreach (classes, it, c) {
+		RBinSymbol *sym = R_NEW0 (RBinSymbol);
+		if (!sym)
+			continue;
+
+		sym->name = r_bin_name_new_from (r_str_newf ("%s::gMetaClass", c->name));
+		sym->vaddr = c->meta_va;
+		sym->paddr = c->meta_va - kext->pa2va_exec;
+		sym->size = 8;
+		sym->forwarder = "NONE";
+		sym->bind = "GLOBAL";
+		sym->type = "OBJECT";
+		sym->ordinal = RVecRBinSymbol_length (symbols);
+		RVecRBinSymbol_push_back (symbols, sym);
+	}
+
+	r_list_free (classes);
+}
+
+static void r_iokit_class_free(void *_c) {
+	RIOKitClass *c = _c;
+	if (c) {
+		free (c->name);
+		free (c);
+	}
+}
+
+static bool try_read_exec_va(RBinFile *bf, RKext *kext, ut64 va, void *dst, size_t len) {
+	return r_buf_read_at (bf->buf, va - kext->pa2va_exec, dst, len) == len;
+}
+
+static bool try_read_printable_cstr(RBinFile *bf, RKext *kext, ut64 va, char **cstr) {
+	*cstr = NULL;
+
+	ut8 buf[128];
+	if (!try_read_exec_va (bf, kext, va, buf, sizeof (buf)))
+		return false;
+
+	int i = 0;
+	for (; i < sizeof (buf); i++) {
+		if (!buf[i])
+			break;
+		if (buf[i] < 0x20 || buf[i] > 0x7e)
+			return false;
+	}
+	if (i < 1 || i >= (int) sizeof (buf))
+		return false;
+
+	char *s = malloc (i + 1);
+	if (!s)
+		return false;
+	memcpy(s, buf, i);
```

Comment:
you can use r_mem_dup, which does malloc+memcpy. for the null byte at the end, maybe just r_str_ndup